File: BE/lhdmir/투 포인터/Gold/14719_빗물.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int h = Integer.parseInt(st.nextToken());
        int w = Integer.parseInt(st.nextToken());

       st = new StringTokenizer(br.readLine());

       int[] arr = new int[w];

       for (int i = 0; i < w; i++) {
           arr[i] = Integer.parseInt(st.nextToken());
       }

       // 빗물 총량
       int volume = 0;
       int left = 0;
       int right = w - 1;
       // 초기 leftMax 값을 arr[left]값, 제일 왼쪽 높이로 설정
       int leftMax = arr[left];
       // 초기 rightMax 값을 arr[right]값, 제일 오른쪽 높이로 설정
       int rightMax = arr[right];

       // left포인터가 right포인터 위치로 이동할때까지 반복
       while (left < right) {
           // 반복문 시작할때마다 최대높이와 현재높이를 비교해서
           // leftMax, rightMax 값을 업데이트
           leftMax = Math.max(leftMax, arr[left]);
           rightMax = Math.max(rightMax, arr[right]);

           // 더 높은쪽을 향해 포인터 이동
           if(leftMax <= rightMax) {
               // leftMax = 3
               // arr[left] = 0 일떄
               // volume = 3
               volume += leftMax  - arr[left++];
           }
           else {
               // right도 left와 같이 계산
               volume += rightMax - arr[right--];
           }
       }

        System.out.println(volume);
    }
}

File: FE/Seo-garden/투포인터_슬라이딩윈도우/Gold/14719_빗물.swift
import Foundation

let HW = readLine()!.split(separator: " ").map { Int($0)! }
let H = HW[0], W = HW[1]
let heights = readLine()!.split(separator: " ").map { Int($0)! }

// 두 포인터 초기화
var left = 0
var right = W - 1
var leftMax = heights[left]
var rightMax = heights[right]
var totalWater = 0

while left < right {        // 두 포인터가 만나기 전까지 반복
    if heights[left] < heights[right] {     // 왼쪽 높이가 오른쪽 높이보다 작은 경우
        left += 1  // 왼쪽 포인터를 오른쪽으로 이동
        leftMax = max(leftMax, heights[left])  // 현재 왼쪽 포인터 위치에서의 최대 높이 갱신
        totalWater += max(0, leftMax - heights[left])  // 현재 위치에서 고일 수 있는 물의 양 계산 후 추가
    } else {  // 오른쪽 높이가 왼쪽 높이보다 작거나 같은 경우
        right -= 1  // 오른쪽 포인터를 왼쪽으로 이동
        rightMax = max(rightMax, heights[right])  // 현재 오른쪽 포인터 위치에서의 최대 높이 갱신
        totalWater += max(0, rightMax - heights[right])  // 현재 위치에서 고일 수 있는 물의 양 계산 후 추가
    }
}

print(totalWater)
